libfreerdp/core/gateway/rdg.c

```diff
@@ -323,34 +359,313 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
 	return TRUE;
 }
 
+static BOOL rdg_write_websocket(rdpRdg* rdg, wStream* sPacket)
+{
+	size_t s;
+	int status;
+	wStream* sWS;
+
+	uint32_t maskingKey;
+
+	int streamPos;
+
+	s = Stream_Length(sPacket);
+	Stream_SetPosition(sPacket, 0);
+
+	if (s > INT_MAX)
+		return FALSE;
+
+	if (s < 126)
+		sWS = Stream_New(NULL, s + 6); // 2 byte "mini header" + 4 byte masking key
+	else if (s < 0x10000)
+		sWS = Stream_New(NULL, s + 8); // 2 byte "mini header" + 2 byte length + 4 byte masking key
+	else
+		sWS = Stream_New(NULL, s + 14); // 2 byte "mini header" + 8 byte length + 4 byte masking key
+
+	if (!sWS)
+		return FALSE;
+
+	winpr_RAND((BYTE*)&maskingKey, 4);
+
+	Stream_Write_UINT8(sWS, WEBSOCKET_FIN_BIT | WebsocketBinaryOpcode);
+	if (s < 126)
+		Stream_Write_UINT8(sWS, s | WEBSOCKET_MASK_BIT);
+	else if (s < 0x10000)
+	{
+		Stream_Write_UINT8(sWS, 126 | WEBSOCKET_MASK_BIT);
+		Stream_Write_UINT16_BE(sWS, s);
+	}
+	else
+	{
+		Stream_Write_UINT8(sWS, 127 | WEBSOCKET_MASK_BIT);
+		Stream_Write_UINT32_BE(sWS, 0); // payload is limited to INT_MAX
+		Stream_Write_UINT32_BE(sWS, s);
+	}
+	Stream_Write_UINT32(sWS, maskingKey);
+
+	// mask as much as possible with 32bit access
+	for (streamPos = 0; streamPos + 4 <= s; streamPos += 4)
```

Comment:
Doing the comment once just here, but we avoid C++ comments because some older compilers don't like them (so if you could drop them in the rest of the code)